3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/PiiPseudonymizerEnrichment.scala

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright (c) 2017 Snowplow Analytics Ltd. All rights reserved.
+ *
+ * This program is licensed to you under the Apache License Version 2.0,
+ * and you may not use this file except in compliance with the Apache License Version 2.0.
+ * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the Apache License Version 2.0 is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+ */
+
+package com.snowplowanalytics
+package snowplow.enrich
+package common
+package enrichments
+
+//Snowplow
+import com.snowplowanalytics.iglu.client.{SchemaCriterion, SchemaKey}
+import com.snowplowanalytics.snowplow.enrich.common.enrichments.registry.{Enrichment, ParseableEnrichment}
+import com.snowplowanalytics.snowplow.enrich.common.utils.MapTransformer.TransformMap
+import com.snowplowanalytics.snowplow.enrich.common.utils.ScalazJson4sUtils
+import org.apache.maven.artifact.versioning.DefaultArtifactVersion
+
+//Scala libraries
+import org.json4s.JsonAST._
+import org.json4s.jackson.JsonMethods.{compact, parse, render}
+import org.json4s.jackson.JsonMethods
+
+//Java
+import java.security.NoSuchAlgorithmException
+import com.github.fge.jsonschema.core.report.ProcessingMessage
+import com.jayway.jsonpath.spi.json.JacksonJsonNodeJsonProvider
+import com.jayway.jsonpath.{JsonPath => JJsonPath, Configuration}
+import com.fasterxml.jackson.databind.JsonNode
+import com.fasterxml.jackson.databind.node.{ObjectNode, TextNode}
+
+
+//scalaz
+import scalaz.Scalaz._
+import scalaz._
+
+object PiiPseudonymizerEnrichment extends ParseableEnrichment {
+  val supportedSchema = SchemaCriterion("com.snowplowanalytics.snowplow", "pii_pseudonymize_config", "jsonscehma", 1, 0, 0)
+
+  def parse(config: JValue, schmaKey: SchemaKey): ValidatedNelMessage[PiiPseudonymizerEnrichment] = {
+    isParseable(config, schmaKey).flatMap(conf => {
+      ScalazJson4sUtils.extract[Boolean](conf, "enabled").flatMap({
+        case enabled => ScalazJson4sUtils.extract[List[JObject]](conf, "parameters", "pii").flatMap({
+          case piiFields => ScalazJson4sUtils.extract[String](config, "parameters", "strategy", "pseudonymize", "hashFunction").flatMap({
```

Comment:
can we find a way around those 4 nested flatMaps?

Reply:
why do you need pattern matching if there aren't any alternatives / type checking?